File: C++/17837_새로운게임2.cpp
#include <iostream>
#include <deque>
using namespace std;

struct Pos{
	int x, y;
}; // 위치 나타내기 위함

struct Dir{
	int num, dir;
}; // 해당 말의 번호와 방향 나타내기 위함

void moveWhite(int idx, Pos p, int x, int y, deque<Dir> **chess, Pos *pos);
void moveRed(int idx, Pos p, int x, int y, deque<Dir> **chess, Pos *pos);

int main() {
	int N, K;
	cin >> N >> K;
	
	int **board = new int*[N];
	deque<Dir> **chess = new deque<Dir>*[N]; //해당 위치에 존재하는 말들을 담을 deque 배열
	for(int i = 0; i < N; i++){
		board[i] = new int[N];
		chess[i] = new deque<Dir>[N];
	}
	
	for(int i = 0; i < N; i++){
		for(int j = 0; j < N; j++){
			cin >> board[i][j];
		}
	}
	
	Pos *pos = new Pos[K]; // K개의 말의 위치 저장
	for(int k = 0; k < K; k++){
		int x, y, dir;
		cin >> x >> y >> dir;
		chess[x-1][y-1].push_back({k,dir-1});
		pos[k] = {x-1,y-1};
	}
	
	int dir[4][2] = {{0,1},{0,-1},{-1,0},{1,0}}; // 1,2,3,4 방향
	int r_dir[4] = {1,0,3,2}; // 방향이 반대로 바뀔 때의 인덱스
	
	int turn = 0; // 수행한 턴의 수를 나타낼 변수
	while(turn++ < 1000){ // 1000번 이상 진행되면 더이상 진행하지 
		for(int k = 0; k < K; k++){ // 말을 순서대로 이동
			Pos p = pos[k];
			int idx = 0; 
			for(idx; idx < chess[p.x][p.y].size(); idx++){ // 말이 현재 위치에서 몇번째 순서에 위치하는지 알아야 함
				if(chess[p.x][p.y][idx].num == k){
					break;
				}
			}
			int x = p.x+dir[chess[p.x][p.y][idx].dir][0], y = p.y+dir[chess[p.x][p.y][idx].dir][1]; // 현재 말이 이동해야 하는 방향으로 이동한 좌표
			if(x >= 0 && x < N && y >= 0 && y < N){ // 범위 내에 있을 때
				if(board[x][y] == 0){ // 흰색으로 이동하려고 할 때
					moveWhite(idx,p,x,y,chess,pos);
					if(chess[x][y].size() >= 4){ // 종료조건: 4개 이상의 말이 모여있을 때
						cout << turn; // 턴 출력 후 종료
						return 0;
					}
				}
				else if(board[x][y] == 1){ 빨간색으로 이동하려고 할 때
					moveRed(idx,p,x,y,chess,pos);
					if(chess[x][y].size() >= 4){ // 종료조건: 4개 이상의 말이 모여있을 때
						cout << turn; // 턴 출력 후 종료
						return 0;
					}
				}
				else if(board[x][y] == 2){ 파란색으로 이동하려고 할 때
					chess[p.x][p.y][idx].dir = r_dir[chess[p.x][p.y][idx].dir]; // 현재 말의 방향을 반대로 바꿔줘야 함(이전에 선언해놓은 배열 이용)
					x = p.x+dir[chess[p.x][p.y][idx].dir][0], y = p.y+dir[chess[p.x][p.y][idx].dir][1]; // 새롭게 이동할 좌표
					if(x >= 0 && x < N && y >= 0 && y < N){ // 범위 내에 있을 때
						if(board[x][y] == 0){ // 흰색으로 이동하려고 할 때
							moveWhite(idx,p,x,y,chess,pos);
						}
						else if(board[x][y] == 1){ // 빨간색으로 이동하려 할 때
							moveRed(idx,p,x,y,chess,pos);
						}
						if(chess[x][y].size() >= 4){ // 종료조건: 4개 이상의 말이 모여있을 때
							cout << turn; // 턴 출력 후 종료
							return 0;
						} // 파란색을 만난 후, 파란색을 또 만났거나 범위를 벗어난 경우, 아무것도 수행하지 않음
					}
				}
			}
			else{ // 범위를 벗어난 경우(파란색과 동일)
				chess[p.x][p.y][idx].dir = r_dir[chess[p.x][p.y][idx].dir]; // 현재 말의 방향을 반대로 바꿔줘야 함(이전에 선언해놓은 배열 이용)
				x = p.x+dir[chess[p.x][p.y][idx].dir][0], y = p.y+dir[chess[p.x][p.y][idx].dir][1]; // 새롭게 이동할 좌표
				if(x >= 0 && x < N && y >= 0 && y < N){ // 범위 내에 있을 때
					if(board[x][y] == 0){ // 흰색으로 이동하려고 할 때
						moveWhite(idx,p,x,y,chess,pos);
					}
					else if(board[x][y] == 1){ // 빨간색으로 이동하려 할 때
						moveRed(idx,p,x,y,chess,pos);
					}
					if(chess[x][y].size() >= 4){ // 종료조건: 4개 이상의 말이 모여있을 때
						cout << turn; // 턴 출력 후 종료
						return 0;
					} // 범위를 벗어나 방향을 바꾼 후, 파란색을 만났거나 범위를 또 벗어난 경우, 아무것도 수행하지 않음
				}
			}
		}
	}
	
	if(turn > 1000){ // 1000번 수행했는데도 게임이 종료되지 않았으면 
		turn = -1; // -1 출력
	}
	
	cout << turn;
	
	return 0;
}

void moveWhite(int idx, Pos p, int x, int y, deque<Dir> **chess, Pos *pos){ // 흰색으로 이동할 때
	deque<Dir> tmp; 
	int c_size = chess[p.x][p.y].size();
	for(idx; idx < c_size; idx++){ // 순서를 그대로 유지해야 하므로 뒤쪽부터 해당 말까지 거꾸로 뽑은 후,
		Dir dir = chess[p.x][p.y].back();
		chess[p.x][p.y].pop_back();
		tmp.push_back(dir);
	}
	
	while(!tmp.empty()){ // 다시 앞쪽말부터 쌓아줌(뒤쪽부터 뽑을 때, push_back으로 이어 붙였으므로 앞쪽말부터 쌓으려면 pop_back하여 쌓아주어야 함)
		Dir dir = tmp.back();
		tmp.pop_back();
		chess[x][y].push_back(dir);
		pos[dir.num] = {x,y}; // 이동된 위치값으로 
	}
}

void moveRed(int idx, Pos p, int x, int y, deque<Dir> **chess, Pos *pos){ // 빨간색으로 이동할 때
	int c_size = chess[p.x][p.y].size();
	for(idx; idx < c_size; idx++){ // 순서를 뒤집어야 하므로 뒤쪽부터 해당 말까지 거꾸로 뽑아서 이동할 위치에 말 쌓아줌
		Dir dir = chess[p.x][p.y].back();
		chess[p.x][p.y].pop_back();
		chess[x][y].push_back(dir);
		pos[dir.num] = {x,y}; // 이동된 위치값으로 변경
	}
}

File: Python/17837_새로운게임2.py
def moveWhite(chess,mal,s,e,num): # 흰색으로 이동할 경우
	move = [] # 이동시켜야 하는 말들을 담은 배열
	while True: # 이동시키려는 말 위에 있는 모든 말들을 순서 변화 없이 이동
		tmp = chess[s[0]][s[1]].pop()
		
		move.append(tmp) # 위에 있는 말 move 배열에 추가
		if tmp == num: # 이동시키려는 말을 만날 때까지
			break
	
	while move: # move 배열에 이동시키려는 말이 위->아래로 정렬되어 있으므로 뒤쪽에 있는 말부터 뽑아서 이동시켜주기 
		tmp = move.pop()
		chess[e[0]][e[1]].append(tmp)
		mal[tmp][0] = e[0]
		mal[tmp][1] = e[1] # 이동시킨 말의 위치 바꿔주기
		
	return mal
	
def moveRed(chess,mal,s,e,num): # 빨간색으로 이동할 경우
	while True: # 이동시키려는 말 위에 있는 모든 말들을 이동시키는데 순서 뒤집기(뒤집으려면 위에 있는 말들이 먼저 이동해야 함)
		tmp = chess[s[0]][s[1]].pop()
		
		chess[e[0]][e[1]].append(tmp) # 위쪽 말부터 목적지로 이동시키기
		
		mal[tmp][0] = e[0]
		mal[tmp][1] = e[1] # 이동시킨 말의 위치 바꿔주기
		
		if tmp == num: # 이동시키려는 말을 만날 때까지
			break

	return mal

if __name__ == '__main__':
	N, K = list(map(int,input().split()))
	board = [list(map(int,input().split())) for _ in range(N)] # 체스판의 색깔을 담고 있는 배열

	mal = []
	chess = [] # 체스판 해당 위치에 몇번 말들이 있는지 표시하기 위한 배열
	for i in range(N):
		tmp = []
		for j in range(N):
			tmp.append([])
		chess.append(tmp)
		
	for k in range(K):
		x, y, dir = list(map(int,input().split()))
		x -= 1
		y -= 1
		chess[x][y].append(k)
		mal.append([x,y,dir-1]) # 말의 위치, 방향 담은 배열
	
	dir = [[0,1],[0,-1],[-1,0],[1,0]] # 우, 좌, 상, 하
	r_dir = [1,0,3,2] # 현재 방향을 반대로 뒤집어야 할 때 어떤 방향으로 뒤집어야할 지 기록해 놓은 배열
	turn = 0
	flag = False # 게임이 정상 종료 되었는지 여부
	while True:
		turn += 1
		
		if turn > 1000:
			break
		
		for i in range(K): # 말 순서대로 이동시키기
			p = mal[i]
      dx = p[0] + dir[p[2]][0]
			dy = p[1] + dir[p[2]][1] # 이동해야할 말의 위치
			if dx < 0 or dx >= N or dy < 0 or dy >= N or board[dx][dy] == 2: # 범위를 벗어났거나 파란색 체스판인 경우
				mal[i][2] = r_dir[mal[i][2]] # 방향 바꾸고
        
        p = mal[i]
        dx = p[0] + dir[p[2]][0]
        dy = p[1] + dir[p[2]][1] # 이동해야할 말의 위치 다시 설정
			
			if dx >= 0 and dx < N and dy >= 0 and dy < N: # 범위 내에 존재하고
				if board[dx][dy] == 0: # 이동하려는 체스판이 흰색일 경우
					mal = moveWhite(chess,mal,[p[0],p[1]],[dx,dy],i)
				elif board[dx][dy] == 1: # 이동하려는 체스판이 빨간색일 경우
					mal = moveRed(chess,mal,[p[0],p[1]],[dx,dy],i)
        # 이동하려는 체스판이 파란색이거나 범위 밖일 경우에는 아무것도 수행하지 않음(만약 방향을 바꿔야하는 경우였다면 위쪽 조건문에서 이미 실행됐을 것임)
					
				if len(chess[dx][dy]) >= 4: # 4개 이상의 말이 한 위치에 쌓여있다면
					flag = True # 정상 종료
					break
		if flag:
			break
		
	if flag: # 정상 종료라면
		print(turn) # 해당 턴 출력
	else: 비정상 종료라면
		print(-1) # -1 출력
